packages/arb-avm-cpp/data_storage/src/arbcore.cpp

```diff
@@ -2119,25 +2121,32 @@ ValueResult<std::unique_ptr<Machine>> ArbCore::getMachineForSideload(
                     std::make_unique<Machine>(*it->second)};
         }
     }
-    // Not found in cache, try the DB
-    ReadSnapshotTransaction tx(data_storage);
-    auto position_res = getSideloadPosition(tx, block_number);
-    if (!position_res.status.ok()) {
-        return {position_res.status, std::unique_ptr<Machine>(nullptr)};
-    }
 
-    auto closest_checkpoint =
-        getClosestExecutionMachine(tx, position_res.data, true);
-    if (std::holds_alternative<rocksdb::Status>(closest_checkpoint)) {
-        return {std::get<rocksdb::Status>(closest_checkpoint), nullptr};
-    }
+    uint256_t gas_target;
+    std::unique_ptr<ExecutionCursor> execution_cursor;
+    {
+        // Not found in cache, try the DB
+        ReadSnapshotTransaction tx(data_storage);
+        auto position_res = getSideloadPosition(tx, block_number);
+        if (!position_res.status.ok()) {
+            return {position_res.status, std::unique_ptr<Machine>(nullptr)};
+        }
+
+        auto closest_checkpoint =
+            getClosestExecutionMachine(tx, position_res.data, true);
+        if (std::holds_alternative<rocksdb::Status>(closest_checkpoint)) {
+            return {std::get<rocksdb::Status>(closest_checkpoint), nullptr};
+        }
 
-    auto execution_cursor = std::make_unique<ExecutionCursor>(
-        std::get<MachineStateKeys>(closest_checkpoint));
+        gas_target = position_res.data;
+        execution_cursor = std::make_unique<ExecutionCursor>(
+            std::get<MachineStateKeys>(closest_checkpoint));
+    }
 
-    auto status = advanceExecutionCursorImpl(
-        tx, *execution_cursor, position_res.data, false, 10, cache, false);
+    auto status = advanceExecutionCursorImpl(*execution_cursor, gas_target,
+                                             false, 10, cache, false);
```

Comment:
Isn't the assumption here that we don't need to advance the execution cursor, and advancing it often wouldn't work because of the staged message?

Reply:
I don't think we want to assume that we'll always have a checkpoint for every sideload. It's true currently, but we may want to adjust our strategy over time and this should work efficiently in the case where a checkpoint exists

Reply:
Hmm... I guess the staged message restriction only applies to the latest sideload, so this works.